[ENH] Marker genes widget redesign

PrimozGodec opened the pull request:
##### Issue
<!-- E.g. Fixes #1, Implements #2, etc. -->
<!-- Or a short description, if the issue does not exist. -->
#180 

##### Description of changes
Redesign of the marker genes widget. 

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

PrimozGodec commented:
The widget is ready now. @pavlin-policar, @BlazZupan can you check if this is what we want to have?

pavlin-policar commented:
So this seems pretty good, but there are a couple of bugs I've found:
1. The filtering is weird. I open the widget and type "epsilon" in the search bar, and I also get _Alpha cells_ with gene _PLCE1_. This happens because the function of _PLCE1_ is _phospholipase C epsilon 1_. I think this shouldn't appear here, and maybe the function shouldn't be considered when the _Group by_ is set to _Cell Type_. Similarly, if I select _Group By_ to be function and type _Alpha_ in the search box, I get genes that don't have alpha anywhere but are markers for _Alpha cells_.
2. I kinda wish that the collapse and expand cell types were synchronized between all and selected genes. So if I collapse a cell type in the left pane, it also gets collapsed in the right pane.
3. I also kinda wish the filtering was applied to the right pane as well. If I have a lot of genes selected and want to remove a single one, that might be annoying to find.
4. If I open the widget and select all the acinar markers, I only have the acinar group on the right. Then I can close the widget and delete it from the workflow. I then reopen the widget. It correctly remembers the selection, but now it's thrown in all the cell types that have a particular gene to the selected pane. An easy way to see this is to type in _GHRL_ and select the epsilon marker gene. Next, delete the widget from the workflow and readd it. Open the widget. In addition to epsilon cells in the selected pane, there are also _Enteroendocrine cells_. 
5. Apparently, genes are not being sent out properly, since I can't connect this to the _Score Genes_ widget.

There are a bunch of other minor convenience things that might be nice to have, but those might require a lot of work. This is already much better than the previous widget, so we should try to get this merged ASAP.

PrimozGodec commented:
@pavlin-policar thank you for the testing and review. I addressed all the suggestions above.

1. Filtering now skips Function or Cell Type (one that is not selected in a group by). Anyway, I now included Entrez ID in filtering. My thoughts are that Entrez ID is so much different (always a number) that it does not affect when some is typing a word that is a gene name or something and if one decides to type number he/she do it by purpose (want to find specific Entrez ID).

2. Collapsing is now synchronized and also persists after drag/drop.

3. Filtering is applied to both panes.

4. Now context should restore variables correctly. Didn't know that names repeat under different cell types.

5. Genes were missing some attributes that were added by the old widget. It is fixed now. Anyway, those genes are not compatible with Score Genes widget it expects gene expressions. Now genes are compatible for example with Gene Sets widget.

Maybe pack other things that need to be changed in a separate issue and can be fixed after the merge.

PrimozGodec commented:
One of the things that will be change is definitely running the widget setup in a thread with progress bar. I will not include that in this PR since it will delay the merge.

PrimozGodec commented:
Tests on stable will keep failing until Orange gets released.

JakaKokosar commented on orangecontrib/bioinformatics/widgets/OWMarkerGenes.py:
```diff
@@ -1,120 +1,735 @@
-    def data_length(self):
-        return len(self._data)
 
+class MarkerGroupContextHandler(ContextHandler):
```
this can be moved to widgets.utils.settings module

JakaKokosar commented on orangecontrib/bioinformatics/widgets/OWMarkerGenes.py:
```diff
@@ -233,222 +890,245 @@ def data(self, value: Table):
+        rows = self.selected_markers_view.model().sourceModel().rootItem.get_data_rows()
+        if len(rows) > 0:
+            metas = [r.metas for r in rows]
+            data = Table(self.data.domain, np.empty((len(metas), 0)), metas=metas)
```
Calls to Table constructor are deprecated: https://github.com/biolab/orange3/commit/76846e2817e36ea24966344e1de27a2be7310a79
